datafusion/expr-common/src/type_coercion/binary.rs

```diff
@@ -1607,6 +1629,18 @@ mod tests {
         }};
     }
 
+    /// Test coercion rules for assymetric binary operators
+    ///
+    /// Applies coercion rules for `$LHS_TYPE $OP $RHS_TYPE` and asserts that the
+    /// the result types are `$RESULT_TYPE1` and `$RESULT_TYPE2` respectively
+    macro_rules! test_coercion_assymetric_binary_rule {
+        ($LHS_TYPE:expr, $RHS_TYPE:expr, $OP:expr, $RESULT_TYPE1:expr, $RESULT_TYPE2:expr) => {{
+            let (lhs, rhs) = get_input_types(&$LHS_TYPE, &$OP, &$RHS_TYPE)?;
+            assert_eq!(lhs, $RESULT_TYPE1);
+            assert_eq!(rhs, $RESULT_TYPE2);
```

Comment:
for `+` operator symmetry is expected, so the assert function could do this for us

Reply:
The operator is symmetric, but how we cast the left and right types is not symmetric. Hence I needed an assymetric test for this.

Reply:
You could have this

```rust
let (a, b) = get_input_types(&$LHS_TYPE, &$OP, &$RHS_TYPE)?;
assert_eq!(a, $RESULT_TYPE1);
assert_eq!(b, $RESULT_TYPE2);

// test symmetry
let (b, a) = get_input_types(&$RHS_TYPE, &$OP, &$LHS_TYPE)?;
assert_eq!(a, $RESULT_TYPE1);
assert_eq!(b, $RESULT_TYPE2);
```